### PR TITLE
Add median-thresholded colocalization algorithm

### LINKED ISSUE #530: Implement median-filtered cosine colocalization algorithm

### PR DESCRIPTION
**Pull request by LachlanStuart**
#### New Feature popup
image

> New Colocalization Algorithm
> METASPACE has an improved colocalization algorithm for ion images!
>
> The median-thresholded cosine distance gives better results than the plain cosine distance when compared to expert judgements. See [Ovchinnikova et al. (2020) ColocML](https://doi.org/10.1093/bioinformatics/btaa085) for more information.
> 
> This new algorithm is now used by default for ranking colocalized annotations. If you wish to go back to the old behavior, cosine distance can be selected in this menu.

#### Colocalization Algorithm popup

image

> The median-thresholded cosine distance algorithm is the best choice for exploring most datasets in METASPACE. It was found to highly reproduce expert rankings of mass spectrometry ion images and performed better than other traditional colocalization measures. 
> Its description and evaluation can be found in: [Ovchinnikova et al. (2020) ColocML.](https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btaa085/5734647)
>
> The cosine distance was previously used as the default colocalization measure. It is still available and can be selected from the list.

## Python side
Please double-check that I've faithfully reproduced the reference implementation's "cosine" output: https://github.com/metaspace2020/coloc/blob/master/measures/NoLearning/ion_intensity_coloc_measures.ipynb

As far as I can tell `sklearn.metrics.pairwise.pairwise_kernels(metric='cosine')` is an acceptable substitute for `gensim.similarities.Similarity` as it generates virtually identical output. I used this script to confirm it:
```python
from sklearn.metrics.pairwise import pairwise_kernels
from gensim.similarities import Similarity
pixels = np.random.rand(1000, 1000)
pixels[pixels < 0.5] = 0 # Make them sparse, because if the two algorithms differed, it would most likely be in the way that sparse elements are handled
docs = [list(zip(px.nonzero()[0], px[px.nonzero()[0]])) for px in pixels]
scipy_result = pairwise_kernels(pixels, metric='cosine')
gensim_result = Similarity(None, docs, pixels.shape[0])[docs]
error = np.abs(scipy_result - gensim_result)
print(f'Max error: {np.max(error)}') # Output: Max error: 3.5762786954052217e-07
```

To minimize time/complexity, clustering is only done with the new algorithm, and the same results are saved for both algorithms. I checked the quality of the results - I think it was slightly improved, but that could just be the effects of a small sample size. 

This needs a two-step deployment - first deploy the sm-engine code, then run `do_colocalization` to recalculate it for all datasets, then do a second deployment with the graphql/ansible config updated to show the new algorithm. As before, I'll run `do_colocalization` on staging to minimize load on prod for the week or so that it takes.

Resolves #530

### CODE REVIEW

**Comment by LachlanStuart** [comment]
I've updated the screenshots & text with changes from feedback

File: metaspace/webapp/src/modules/Annotations/annotation-widgets/default/RelatedAnnotations.vue [hunk 1]
```diff
@@ -78,6 +78,11 @@
     >
       No annotations found.
     </p>
+
+    <div
+      v-if="query === 'colocalized'"
```
[the remark this answers is not in the document]
Comment (richardgoater) [reply metaspace/webapp/src/modules/Annotations/annotation-widgets/default/RelatedAnnotations.vue]: this was discussed offline, we said that it's tricky in Vue templates, better suited to tsx.

File: metaspace/webapp/src/modules/Annotations/annotation-widgets/ColocalizationSettings.vue [hunk 2]
```diff
@@ -1,11 +1,40 @@
...
+  <span id="colocalization-settings">
+    <el-form label-position="top">
+      <el-form-item>
+        <span slot="label">
+          Colocalization algorithm
+          <el-popover
+            trigger="hover"
+            placement="top"
+          >
+            <div style="max-width: 550px;">
```
Comment (richardgoater) [inline metaspace/webapp/src/modules/Annotations/annotation-widgets/ColocalizationSettings.vue]: This is fine for now but in future I would like to set max-widths using Tailwind. This allows elements to be sized according to a scale, which is a bit more "Gestalt".

File: metaspace/webapp/src/modules/Annotations/annotation-widgets/ColocalizationSettings.vue [hunk 3]
```diff
@@ -1,11 +1,40 @@
...
+      <el-form-item>
+        <span slot="label">
+          Colocalization algorithm
+          <el-popover
+            trigger="hover"
+            placement="top"
+          >
+            <div style="max-width: 500px;">
+              <p>
+                <b>Median-thresholded cosine distance</b> is the best choice for exploring most biological datasets.
```
[the remark this answers is not in the document]
Comment (LachlanStuart) [reply metaspace/webapp/src/modules/Annotations/annotation-widgets/ColocalizationSettings.vue]: I have no strong opinion in this particular case, but `<b>` is used throughout the codebase, whereas there are only 2 uses of `<strong>` and no uses of `<em>`.

File: metaspace/webapp/src/modules/App/NewFeaturePopup.vue [hunk 4]
```diff
@@ -196,7 +215,7 @@ export default class NewFeaturePopup extends Vue {
   .nf-container {
     background: $popper-background-color;
     color: black;
-    width: 400px;
+    width: 500px;
```
[the remark this answers is not in the document]
Comment (LachlanStuart) [reply metaspace/webapp/src/modules/App/NewFeaturePopup.vue]: I changed this so that the "Ovchinnikova et al. (2020) ColocML" doesn't get split across multiple lines (at least with 1x DPR), as it looks a bit ugly when it does split. 

There's only one other NewFeaturePopup that also uses this class, which is now a bit wider but doesn't look any worse:

image

File: metaspace/webapp/src/modules/Annotations/annotation-widgets/ColocalizationSettings.vue (hunk 2 above)
Comment (LachlanStuart) [reply metaspace/webapp/src/modules/Annotations/annotation-widgets/ColocalizationSettings.vue]: Agreed. I'll gladly join the cause and help clean up the mess of inline `style`s once Tailwind is in place.
